ldes-fragmentisers/ldes-fragmentisers-substring/src/main/java/be/vlaanderen/informatievlaanderen/ldes/server/fragmentisers/substring/SubstringFragmentationStrategy.java

```diff
@@ -5,46 +5,82 @@
-	private final SubstringBucketiser substringBucketiser;
 	private final SubstringFragmentFinder substringFragmentFinder;
 	private final SubstringFragmentCreator substringFragmentCreator;
+	private final SubstringPreProcessor substringPreProcessor;
 
 	public SubstringFragmentationStrategy(FragmentationStrategy fragmentationStrategy,
 			ObservationRegistry observationRegistry,
-			SubstringBucketiser substringBucketiser,
-			SubstringFragmentFinder substringFragmentFinder, SubstringFragmentCreator substringFragmentCreator,
-			TreeRelationsRepository treeRelationsRepository) {
+			SubstringFragmentFinder substringFragmentFinder,
+			SubstringFragmentCreator substringFragmentCreator,
+			TreeRelationsRepository treeRelationsRepository,
+			SubstringPreProcessor substringPreProcessor) {
 		super(fragmentationStrategy, treeRelationsRepository);
 		this.observationRegistry = observationRegistry;
-		this.substringBucketiser = substringBucketiser;
 		this.substringFragmentFinder = substringFragmentFinder;
 		this.substringFragmentCreator = substringFragmentCreator;
+		this.substringPreProcessor = substringPreProcessor;
 	}
 
 	@Override
 	public void addMemberToFragment(LdesFragment parentFragment, Member member, Observation parentObservation) {
-		Observation substringFragmentationObservation = Observation.createNotStarted("substring fragmentation",
-				observationRegistry)
+		final Observation substringFragmentationObservation = startFragmentationObservation(parentObservation);
+
+		final LdesFragment rootFragment = prepareRootFragment(parentFragment);
+		addMemberToFragments(parentFragment, member, substringFragmentationObservation, rootFragment);
+		substringFragmentationObservation.stop();
+	}
+
+	private LdesFragment prepareRootFragment(LdesFragment parentFragment) {
+		final LdesFragment rootFragment = substringFragmentCreator.getOrCreateSubstringFragment(parentFragment,
+				ROOT_SUBSTRING);
+		super.addRelationFromParentToChild(parentFragment, rootFragment);
+		return rootFragment;
+	}
+
+	private void addMemberToFragments(LdesFragment parentFragment,
+			Member member,
+			Observation substringFragmentationObservation,
+			LdesFragment rootFragment) {
+		final String substringTarget = substringPreProcessor.getSubstringTarget(member);
+		final List<String> tokens = substringPreProcessor.tokenize(substringTarget);
+		final Set<String> substringsThatContainMember = new HashSet<>();
+		tokens.forEach(token -> {
+			final List<String> buckets = substringPreProcessor.bucketize(token);
+			if (buckets.stream().noneMatch(substringsThatContainMember::contains)) {
+				final LdesFragment substringFragment = substringFragmentFinder
+						.getOpenOrLastPossibleFragment(parentFragment, rootFragment, buckets);
+				substringsThatContainMember.add(extractSubstringFromFragmentId(substringFragment.getFragmentId()));
+				super.addMemberToFragment(substringFragment, member, substringFragmentationObservation);
```

Comment:
It would be possible to use this method in parallel if you know the all the substringFragments a priori and don't calculate this in the forEach.